File: dijkstra/dijkstra.js
const fs = require('fs');
const readline = require('readline');


class MinHeap {
    constructor() {
        this.heap = [];
    }

    push(val) {
        this.heap.push(val);
        this._bubbleUp();
    }

    pop() {
        if (this.size() === 0) return null;
        if (this.size() === 1) return this.heap.pop();
        
        const min = this.heap[0];
        this.heap[0] = this.heap.pop();
        this._sinkDown();
        return min;
    }

    size() {
        return this.heap.length;
    }

    _bubbleUp() {
        let index = this.heap.length - 1;
        while (index > 0) {
            const parentIndex = Math.floor((index - 1) / 2);
            if (this.heap[index][0] >= this.heap[parentIndex][0]) break;
            
            [this.heap[index], this.heap[parentIndex]] = [this.heap[parentIndex], this.heap[index]];
            index = parentIndex;
        }
    }

    _sinkDown() {
        let index = 0;
        const length = this.heap.length;
        
        while (true) {
            let leftChildIdx = 2 * index + 1;
            let rightChildIdx = 2 * index + 2;
            let swap = null;

            if (leftChildIdx < length) {
                if (this.heap[leftChildIdx][0] < this.heap[index][0]) {
                    swap = leftChildIdx;
                }
            }

            if (rightChildIdx < length) {
                if (
                    (swap === null && this.heap[rightChildIdx][0] < this.heap[index][0]) ||
                    (swap !== null && this.heap[rightChildIdx][0] < this.heap[swap][0])
                ) {
                    swap = rightChildIdx;
                }
            }

            if (swap === null) break;

            [this.heap[index], this.heap[swap]] = [this.heap[swap], this.heap[index]];
            index = swap;
        }
    }
}


const INFINITY = Infinity;

async function loadGraph() {
    const graph = {};

    const fileStream = fs.createReadStream("D:/CODE/OESK/graph_with_weights.txt");

    const rl = readline.createInterface({
        input: fileStream,
        crlfDelay: Infinity
    });

    console.log("Loading graph...");

    for await (const line of rl) {
        const trimmedLine = line.trim();
        
        if (trimmedLine.startsWith("#") || !trimmedLine) {
            continue;
        }
        const parts = trimmedLine.split(/\s+/);

        if (parts.length !== 3) {
            continue;
        }

        const fromNode = Number(parts[0]);
        const toNode = Number(parts[1]);
        const weight = Number(parts[2]);

        if (!graph[fromNode]) graph[fromNode] = {};
        if (!graph[toNode]) graph[toNode] = {}; 

        graph[fromNode][toNode] = weight;
    }

    console.log("Graph loaded.");
    return graph;
}

function dijkstraHeap(G, startNode) {
    const shortestPaths = {};
    const visited = new Set();
    const parents = {};
    
    const heap = new MinHeap();

    for (const node of Object.keys(G)) {
        shortestPaths[node] = INFINITY;
        parents[node] = null;
    }

    shortestPaths[startNode] = 0;
    
    // Heap stores [distance, node]
    heap.push([0, startNode]);

    while (heap.size() > 0) {
        const [currentDistance, currentNode] = heap.pop();

        if (visited.has(Number(currentNode))) {
            continue;
        }
        visited.add(Number(currentNode));

        const neighbors = G[currentNode];
        
        if (neighbors) {
            for (const [neighborStr, weight] of Object.entries(neighbors)) {
                const neighbor = Number(neighborStr);
                const distance = currentDistance + weight;

                if (distance < shortestPaths[neighbor]) {
                    shortestPaths[neighbor] = distance;
                    parents[neighbor] = currentNode;
                    heap.push([distance, neighbor]);
                }
            }
        }
    }

    return { costs: shortestPaths, parents };
}

function reconstructPath(parents, endNode) {
    const path = [];
    let current = endNode;

    if (parents[current] === undefined && current !== undefined) {
        return [];
    }

    while (current !== null && current !== undefined) {
        path.push(current);
        current = parents[current];
    }

    return path.reverse();
}

(async () => {
    try {
        const G = await loadGraph();
        
        const start = 1;
        const end = 425875;

        if (!G[start]) {
            console.error(`Start node ${start} not found in graph.`);
            return;
        }

        console.log(`Running Dijkstra from ${start} to ${end}...`);
        const { costs, parents } = dijkstraHeap(G, start);
        
        const path = reconstructPath(parents, end);

        if (costs[end] === INFINITY) {
            console.log(`No path found between ${start} and ${end}.`);
        } else {
            console.log(`Cost to ${end}: ${costs[end]}`);
            console.log(`Path to ${end}:`, path);
        }

    } catch (error) {
        console.error("An error occurred:", error);
    }
})();

File: dijkstra/dijkstra.py
import heapq
graph_test = {
  "A": {"B": 3, "C": 3},
  "B": {"A": 3, "D": 3.5, "E": 2.8},
  "C": {"A": 3, "E": 2.8, "F": 3.5},
  "D": {"B": 3.5, "E": 3.1, "G": 10},
  "E": {"B": 2.8, "C": 2.8, "D": 3.1, "G": 7},
  "F": {"G": 2.5, "C": 3.5},
  "G": {"F": 2.5, "E": 7, "D": 10},
}

# funkcja do załadowania grafu
def load_graph():
  graph = {}
  # otworzenie pliku
  with open("D:/CODE/OESK/graph_with_weights.txt", "r") as file:
    for line in file:
      # jesli zaczynamy linie od # lub są puste linie to skip
      if line.startswith("#") or not line.strip():
        continue
      # dzielimy każdy node na dwie różne listy
      parts = line.split()
      # jeśli byłby błąd i pojawiłyby się więcej lub mniej niz 3 to skip
      if len(parts) != 3:
        continue
      
      from_node, to_node, weight = parts
      # zmieniamy na inty
      from_node = int(from_node)
      to_node = int(to_node)
      weight = int(weight)

      # jesli klucza nie ma w grafie to dodajemy
      if from_node not in graph:
        graph[from_node] = {}
      # kazda wartosc do klucza dodajemy
      graph[from_node][to_node] = weight
      # jeśli dziecko jest ostatnie to tworzymy je rodzicem z pustym zbiorem
      if to_node not in graph:
        graph[to_node] = {}

  return graph

infinity = float("inf")

def dijkstra_heap(G, start_node):
  shortest_paths = {} 
  visited = set()
  heap = []
  
  parents = {}

  for node in G.keys():
    shortest_paths[node] = infinity
    parents[node] = None 

  shortest_paths[start_node] = 0 
  
  heapq.heappush(heap, (0, start_node))

  while heap:
    (current_distance, current_node) = heapq.heappop(heap)
    
    if current_node in visited:
      continue
    visited.add(current_node)

    for neighbor, weight in G[current_node].items():
      distance = current_distance + weight

      if distance < shortest_paths[neighbor]:
        shortest_paths[neighbor] = distance
        parents[neighbor] = current_node 
        heapq.heappush(heap, (distance, neighbor))

  return shortest_paths, parents

def reconstruct_path(parents, end_node):
  path = []
  current = end_node
  
  while current is not None:
    path.append(current)
    current = parents[current]
  
  return path[::-1]

G = load_graph()
start = 1
end = 425875

costs, parents = dijkstra_heap(G, start)
path = reconstruct_path(parents, end)

print(f"Cost to {end}: {costs[end]}")
print(f"Path to {end}: {path}")
